### ISSUE TITLE
Adding cross-validation to fit in spektral

### ISSUE THREAD

**Issue by tensor-tensor** [danielegrattarola/spektral#275]
Hello,

`validation-data = (X_valid, y_valid)` is a known option to be added to `fit()` of keras. Though, I don't know how one should add a test loader of spektral to this option. In particular, Given the code below,

```
import numpy as np
import matplotlib.pyplot as plt

from tensorflow.keras.losses import CategoricalCrossentropy
from tensorflow.keras.metrics import categorical_accuracy
from tensorflow.keras.optimizers import Adam

from spektral.data import DisjointLoader
from spektral.datasets import TUDataset
from spektral.models import GeneralGNN

np.random.seed(0)

batch_size = 16
learning_rate = 0.0001
epochs = 5

data = TUDataset("PROTEINS")
np.random.shuffle(data)
split = int(0.8 * len(data))
data_tr, data_te = data[:split], data[split:]

loader_tr = DisjointLoader(data_tr, batch_size=batch_size, epochs=epochs)
loader_te = DisjointLoader(data_te, batch_size=batch_size)

model = GeneralGNN(data.n_labels, activation="softmax")
optimizer = Adam(learning_rate)
loss_fn = CategoricalCrossentropy()
model.compile(loss=loss_fn,
              optimizer=optimizer,
              metrics=categorical_accuracy)

history = model.fit(loader_tr.load(),
                    steps_per_epoch=loader_tr.steps_per_epoch,
                    epochs=epochs)
```

Adding `validation_data=loader_te.load()` to `fit()` does not do the job since the execution stops right after the first epoch. 

What is the way to do so?

**Comment by danielegrattarola** [comment]
Hi, 

I realized there are no examples that show how to use validation data, I'll get around to it asap. 

Anyway, the correct way of passing validation data is to create a dedicated loader without specifying the number of epochs (so that the loader will cycle indeterminately)  and pass it as validation data with  the `validation_step` keyword set: 

```python
loader_tr = DisjointLoader(data_tr, batch_size=batch_size, epochs=epochs)
loader_va = DisjointLoader(data_va, batch_size=batch_size)

model.fit(
    loader_tr.load(),
    steps_per_epoch=loader_tr.steps_per_epoch,
    epochs=epochs,
    validation_data=loader_va.load(),
    validation_steps=loader_va.steps_per_epoch,
)
```

Cheers

**Comment by tensor-tensor** [comment]
Thanks for the advice, though I guess something's wrong here. In particular, running

```
import numpy as np
import matplotlib.pyplot as plt

from tensorflow.keras.losses import CategoricalCrossentropy
from tensorflow.keras.metrics import categorical_accuracy
from tensorflow.keras.optimizers import Adam

from spektral.data import DisjointLoader
from spektral.datasets import TUDataset
from spektral.models import GeneralGNN

np.random.seed(0)

batch_size = 16
learning_rate = 0.0001
epochs = 100

data = TUDataset("PROTEINS")
np.random.shuffle(data)
split = int(0.8 * len(data))
data_tr, data_te = data[:split], data[split:]

loader_tr = DisjointLoader(data_tr, batch_size=batch_size, epochs=epochs)
loader_te = DisjointLoader(data_te, batch_size=batch_size)

model = GeneralGNN(data.n_labels, activation="softmax")

v_model = GeneralGNN(data.n_labels, activation="softmax")


optimizer = Adam(learning_rate)
loss_fn = CategoricalCrossentropy()

model.compile(loss=loss_fn,
              optimizer=optimizer,
              metrics=categorical_accuracy)

v_model.compile(loss=loss_fn,
              optimizer=optimizer,
              metrics=categorical_accuracy)

v_history = v_model.fit(loader_tr.load(),
                    steps_per_epoch=loader_tr.steps_per_epoch,
                    epochs=epochs,
                    validation_data=loader_te.load(),
                    validation_steps=loader_te.steps_per_epoch)

plt.plot(v_history.history['loss'])
plt.plot(v_history.history['categorical_accuracy'])
plt.plot(v_history.history['val_loss'])
plt.plot(v_history.history['val_categorical_accuracy'])
plt.xlabel('epoch')
plt.legend(["Loss", "Categorical Accuracy", "Val_Loss", "Val_Categorical Accuracy"])
plt.show()
```
returns

![rrrr](https://user-images.githubusercontent.com/89932608/131827690-1f610271-a7b9-4681-8a5d-7330f45a5681.png)

which is weird as it seems the accuracy after the cross validation is worse than the one before that. I do know that those oscillations are due to the nature of the loaded TU dataset. But how can cross validation make the results worse?

**Comment by danielegrattarola** [comment]
That's just the dataset, which is really small, so your GNN is overfitting the training data and it has poor validation performance. 
You can try using either a simpler network to avoid overfitting or a different dataset.
